### ISSUE TITLE
Beastify does not install

### ISSUE THREAD

**Issue by Ge-Bra** [mdn/webextensions-examples#1]
It is missing the "applications" property in the manifest

**Comment by wbamberg** [comment]
Yes, beastify does not currently work in Firefox due to some bugs, so I omitted the "applications" key (otherwise, since Google Chrome doesn't like the "applications" key, it would not be installable anywhere!). I should have noted this, but I hadn't really expected anyone to stumble on this one. I have now updated the README: https://github.com/mdn/webextensions-examples/blob/master/README.md.

Once the Firefox bugs are fixed, I'll add the "applications" key back and close this bug.

Thanks for trying it out!

**Comment by Mottie** [comment]
The documentation should also explicitly mention that the "id" needs to be in an email format. 

```js
"applications": {
  "gecko": {
    "id": "myextension@example.com"
  }
}
```

I did find a [reference here](https://developer.mozilla.org/en-US/Add-ons/Install_Manifests#id), but it took a lot of link clicking to get there.

**Comment by ncalexan** [comment]
Beastify now installs, but it doesn't work.  I get
```
TypeError: tab is null (ext-tabs.js:442:13)
```
Is this expected?

**Comment by wbamberg** [comment]
@ncalexan 
> Is this expected?

Yes! There are at least 2 issues that stop it from working in current Nightlies. This is noted in my comment upthread and in the docs for these examples: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Examples.

**Comment by Omar-Asumani** [comment]
'tab' problem could easily be fixed if you replace 'currentWindow' key in 'chrome.tabs.query' method with 'lastFocusedWindow'. Also you should add 'tabs' permissions in manifest. Nevertheless I still couldn't successfully run it. I believe there are still problems with obtaining web accessible resources. I am developing extension of my own and I face problems with 'chrome.extension.getURL'. I simply can't find workaround to access extension resources from content.

**Comment by wbamberg** [comment]
> 'tab' problem could easily be fixed if you replace 'currentWindow' key
> in 'chrome.tabs.query' method with 'lastFocusedWindow'.

I don't think so. I think this "tab is null" error is because we were omitting the first argument to "tabs.executeScript" [1]. According to the Chrome docs, if you omit this, you execute the script in the current tab [2]. This does not work on Firefox [3]. The workaround is to pass null instead of just omitting that argument. Since the workaround is a simple one, and the code seems less magical that way anyway, I just updated the example to do that.

However, the next bug is that Firefox doesn't support the "callback" argument to executeScript [4]. I'm using this so I know that the content script has executed (and therefore that message listeners have been added) before I try to send a message to the content script.

Actually, I don't know if this is needed, or if I can just send messages right after calling executeScript. I have tried sending messages right after executeScript, and it works, but I don't know if it just happens to work in my case, or if it is generally a safe thing to do (for example: if the API queues messages until the script has executed. But that sounds kind of unlikely). If it is needed, then it's hard to fix this example without bug 1210583 being fixed, so the example is kind of in limbo.

>  I believe there are still problems with obtaining web accessible resources.

One problem that would affect this example is that wildcards in web_accessible_resources don't work [5]. If you list all the resources explicitly, then it should work. But that would not be a natural way to write an add-on like this, so I don't want to make that change, and would prefer to wait for the bug fix.

>  Also you should add 'tabs' permissions in manifest.

Are you sure about that? It works fine in Chrome as it is, without "tabs". And I don't see any permission problems in Firefox either. I thought specifying a host (or every host, in this case :) ) gave you permission to use executeScript? What I really ought to do here is use "activeTab" instead of requesting access to every host in the world, but I can't, because Firefox does not yet support the activeTab permission[6].

>  I am developing extension of my own and I face problems with 'chrome.extension.getURL'

What problem? It works for me. It didn't work a few weeks ago, but seems to be fixed in current Nightlies.

[1] https://github.com/mdn/webextensions-examples/blob/master/beastify/popup/choose_beast.js#L22 - but note that I've just updated this to pass "null".
[2] https://developer.chrome.com/extensions/tabs#method-executeScript
[3] https://bugzilla.mozilla.org/show_bug.cgi?id=1208257
[4] https://bugzilla.mozilla.org/show_bug.cgi?id=1210583
[5] https://bugzilla.mozilla.org/show_bug.cgi?id=1217944
[6] https://bugzilla.mozilla.org/show_bug.cgi?id=1190688

**Comment by wbamberg** [comment]
OK, so I've had it confirmed that the callback stuff is not needed, so I made the other changes and this add-on ought to work in Firefox Nightly now.
